### ISSUE TITLE
UI gets freezed on api/v1/label/__name__/values?_=1566195610386

### ISSUE THREAD

**Issue by kshitijcode** [prometheus/prometheus#5919]
I tried using the prometheus UI . 

On loading the endpoint, it got stalled executing the request 
api/v1/label/__name__/values?_=1566197849138. The response was 200OK but the browser is freezing. 

We do have a large number of metrics/labels in our cluster.

I saw a previous issue where limit was reduced to 10k but I guess thats not helping. Is there any way to configure this? 

Using prometheus version : 2.10.0

**Comment by simonpasquier** [comment]
How many time series do you have?

**Comment by FUSAKLA** [comment]
This issue was encountered using Thanos which uses the same GUI and theoretically can have unlimited number or label values even when following best practices. Not saying Prometheus UI should be able to handle amount of data this big but option to at least disable the hinting at all would be great possibly?

**Comment by brian-brazil** [comment]
I thought we had something in the javascript to limit this?

**Comment by FUSAKLA** [comment]
It was discussed on Thanos slack before https://cloud-native.slack.com/archives/CK5RSSC10/p1565939498149400

from the discussion:
> Yes, we do have large amount of metrics/labels. And yes the response is 200 OK. Can we disable or restrict /limit the values somehow?
The response is 200OK but it says request is not finished yet. The resource size download is about 125mb and request takes around 24seconds. 

The amount of data in this case is probably really big and JS can limit this only once it is loaded so in cases like this might be good to omit theese request at all? There is already the checkbox for enable/disable query history so the same way the hinting could be disabled maybe?

**Comment by brian-brazil** [comment]
Uhm, how do you have several million metric names?

**Comment by ramshankarjaiswal** [comment]
Current status of our prometheus - monitoring 2262 targets

sum((scrape_samples_scraped)) = 2172660
Output of ({__name__=~".+"}) -
Warning! Fetched 2177672 series, but displaying only first 10000.

Are the above two queries pretty much giving me the same answer? i.e. total current time series? and how do they differ from prometheus_tsdb_head_series < this gives us unique time series over last 2-3 hours?

In the past we've had some applications misbehave and expose 100k to 300k metrics each from about a 10+ servers .. that is why we have so many metric names.

This causes the browser to freeze after few seconds of running any query.. same thing happens when we use thanos for global query view.

Thanks to your book @brian-brazil I've come across https://www.robustperception.io/using-sample_limit-to-avoid-overload which I will now implement to avoid this in future.

**Comment by brian-brazil** [comment]
That query is for metrics, not metric names which is what matters here. Even in a quite large and very well instrumented system you're unlikely to go much beyond 10k metric names.

**Comment by ramshankarjaiswal** [comment]
```
curl http://prometheus:9090/api/v1/label/__name__/values > metricnames.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  121M    0  121M    0     0  1191k      0 --:--:--  0:01:44 --:--:-- 1551k

cat metricnames.json | tr '"' ' ' | tr -d "," | wc -w
2129511
```
We have about 2 million metric names.

**Comment by brian-brazil** [comment]
Likely you've got some labelvalues that someone put in a metric name then.

**Comment by ramshankarjaiswal** [comment]
I understand that we have million more metric names then ideal. Would it be possible to prevent its loading since it freezes the browser?

**Comment by ramshankarjaiswal** [comment]
As a temporary workaround I'm using a chrome extension https://chrome.google.com/webstore/detail/http-request-blocker/eckpjmeijpoipmldfbckahppeonkoeko?hl=en to block this pattern https://myprometheus/api/v1/label/__name__/values?_=*

Although I can't ask other users of my prometheus to do the same. It would be nice if we can disable loading of available metrics through a server side prometheus config.

**Comment by dgl** [comment]
We're also seeing issues with this, with around 100k metrics the UI is just about usable on a high end machine, but noticeably slow. Our setup is also Thanos and then we have a custom Thanos store to bridge our OpenTSDB instance (https://github.com/G-Research/geras).

As a way of reproducing this try opening the JavaScript console on the Prometheus graph UI and running something like:
```js
pageConfig.allMetrics.push(...[...Array(100000).keys()].map(x => "fairly_long_test_"+x))

pageConfig.allMetrics.push(...[...Array(100000).keys()].map(x => "fairly_long_test_"+(100000+x)))
```

It seems there is quite a performance difference between just running the first and adding another 100k metrics. (But this is much better since the changes to limit to 1000 items.)

How about a solution that adds a maximum parameter to the `label/.../values` API? The JavaScript code could then fetch that with a limit around 10k when the graph UI loads -- if the number returned is under 10k then it knows it has all the metrics locally. If it is over then it knows to fetch them from the server -- with some kind of substring parameter to the label values API. The main downside of this is the fuzzy matching wouldn't fully work for users with a lot of metrics, but then the UI would likely be much more responsive. (I'm happy to work on something like this.)

**Comment by brian-brazil** [comment]
There's no need to support users that have 100k metric names, as such a setup is almost certainly due to putting label values in metric names.

**Comment by dgl** [comment]
I understand that's true in an ideal world, we're attempting to make a large OpenTSDB setup compatible with Prometheus. OpenTSDB has far more strict limits on tags (it's version of labels), in particular there is a limit of 8 tags on any metric so that somewhat constrains the use of them.

It seems like with a few simple and unobtrusive changes we can improve this.

**Comment by bwplotka** [comment]
I think adding some limit is a fair request. It would be nice to have in Prometheus anyway, especially if we think that this is the rule: ` no need to support users that have 100k metric names, as such a setup is almost certainly due to putting label values in metric names.` (:  

In the worst case, we can just limit Thanos Querier UI JS, but I think there is no harm in doing this on Prometheus UI as well.

**Comment by brian-brazil** [comment]
I was more making the point that the UI not quite working in such a situation is acceptable.

**Comment by GiedriusS** [comment]
@brian-brazil what do you think about the following: perhaps we should calculate at the beginning how many metrics there are and if it's past a certain threshold then we shouldn't show the dropdown? I feel like it would be hard to implement the same functionality **correctly** that the UI currently has if the label values API would support pagination.

**Comment by brian-brazil** [comment]
I don't think that'd help here, as merely the JS parsing the large JSON result appears to be an issue.

**Comment by GiedriusS** [comment]
Exactly, that's the problem. What I am proposing is for the UI to do a query like `count({__name__=~".+"})` and check if it passes a certain threshold. Obviously, this won't give us the exact numbers and would put unnecessary pressure on the node so maybe it's not the best solution.

What do you think about this other idea that I had, @brian-brazil: what if there was a button in the place of where the drop-down with the list of metrics is at the moment and only after it gets clicked, the list of metrics would be loaded from Prometheus? Optionally, there could a checkbox next to it which would make sure that we always load the metrics but there might not be enough place. I need to play around with this. I feel like this way we would get the best of both worlds. 

Either way, solving this would really help out my old laptop.

**Comment by brian-brazil** [comment]
I don't think adding extra UI elements for the user to worry about is the thing to do. Whatever the result is, it should Just Work.

**Comment by dgl** [comment]
I've made a pull request that addresses some of the issues here; it hides the metric selector if there's more than 5,000 metrics (based on a little bit of testing with Chrome) and also changes the fuzzy matching to not try to match on a huge number of metric names.

**Comment by brian-brazil** [comment]
We looked at this as part of the bug scrub, and this is a dupe of #5421.
